upstairs/src/lib.rs

```diff
@@ -2873,107 +2887,121 @@ impl Downstairs {
-                response.hashes.len(),
-            );
+        if response.block_contexts.is_empty()
+            || response
+                .block_contexts
+                .iter()
+                .all(|ctx| ctx.encryption_context.is_none())
+        {
+            // No block context(s), or no block context contained an encryption
+            // context, in the response!
+            //
+            // Either this is a read of an unwritten block, or an attacker
+            // removed the encryption contexts from the db.
+            //
+            // XXX if it's not a blank block, we may be under attack?
+            assert!(response.data[..].iter().all(|&x| x == 0));
+            return Ok(None);
+        }
 
-            for (i, ctx) in encryption_context_iter {
-                // Validate integrity hash before decryption
-                let computed_hash = integrity_hash(&[
-                    &ctx.nonce[..],
-                    &ctx.tag[..],
-                    &response.data[..],
-                ]);
+        let mut valid_hash = None;
 
-                if computed_hash == response.hashes[i] {
-                    successful_hash = true;
-                    vh = Some(computed_hash);
+        let mut successful_decryption = false;
+        let mut successful_hash = false;
 
-                    // Now that the integrity hash was verified, attempt
-                    // decryption.
-                    //
-                    // Note: decrypt_in_place does not overwrite the buffer if
-                    // it fails, otherwise we would need to copy here. There's a
-                    // unit test to validate this behaviour.
-                    let decryption_result = encryption_context
-                        .decrypt_in_place(
-                            &mut response.data[..],
-                            Nonce::from_slice(&ctx.nonce[..]),
-                            Tag::from_slice(&ctx.tag[..]),
-                        );
+        // Attempt decryption with each encryption context, and fail if all
+        // do not work. The most recent encryption context will most likely
+        // be the correct one so start there.
+        for ctx in response.block_contexts.iter().rev() {
+            let block_encryption_ctx =
+                if let Some(block_encryption_ctx) = &ctx.encryption_context {
+                    block_encryption_ctx
+                } else {
+                    // this block context is missing an encryption context!
+                    // continue to see if another block context has a valid one.
+                    continue;
+                };
+
+            // Validate integrity hash before decryption
+            let computed_hash = integrity_hash(&[
+                &block_encryption_ctx.nonce[..],
+                &block_encryption_ctx.tag[..],
+                &response.data[..],
+            ]);
+
+            if computed_hash == ctx.hash {
+                successful_hash = true;
```

Comment:
No matter what happens below this, we are exiting the `for` loop here.

Could there be a situation where the `computed_hash == ctx.hansh`, but the decryption failed, and there were other `block_contexts` that we did not get to yet that did pass?

Reply:
I think so, yes, if there are two block contexts with difference nonces and tags, but that both return the same `integrity_hash`:

```rust
let hash1 = integrity_hash(&[&ctx1.nonce[..], &ctx1.tag[..], &response.data[..]]);
let hash2 = integrity_hash(&[&ctx2.nonce[..], &ctx2.tag[..], &response.data[..]]);

hash1 == hash2
```

Given that xxHash is a non-cryptographic hash and that the hash value is a `u64`, collisions are possible, though unlikely (I've been running something to try and find such a collision for several hours). The fix for this is to **not** break if decryption is unsuccessful, and try the other block contexts. I've done this in 8a8bb8a

Reply:
Ah, I had not considered the collision case, but yeah, that could be possible.